golang/cosmos/vm/action.go

```diff
@@ -0,0 +1,104 @@
+package vm
+
+import (
+	"reflect"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var (
+	zeroTime         = time.Time{}
+	actionHeaderType = reflect.TypeOf(ActionHeader{})
+)
+
+// Jsonable is a value, j, that can be passed through json.Marshal(j).
+type Jsonable interface{}
+
+// ActionHeader must be embedded in all actions.  It is populated by PopulateAction.
+type ActionHeader struct {
+	// Type defaults to the `default:"..."` tag of the embedder's ActionHeader field.
+	Type string `json:"type"`
+	// BlockHeight defaults to sdk.Context.BlockHeight().
+	BlockHeight int64 `json:"blockHeight"`
+	// BlockTime defaults to sdk.Context.BlockTime().Unix().
+	BlockTime int64 `json:"blockTime"`
+}
+
+// ActionPusher enqueues data for later consumption by the controller.
+type ActionPusher func(ctx sdk.Context, action Jsonable) error
+
+func setActionHeaderDefaults(ctx sdk.Context, defaultType string, ah *ActionHeader) {
+	// Default the Type field to the default tag.
+	if len(defaultType) != 0 && len(ah.Type) == 0 {
+		ah.Type = defaultType
+	}
+
+	// Default the block height.
+	if ah.BlockHeight == 0 {
+		ah.BlockHeight = ctx.BlockHeight()
+	}
+
+	if ah.BlockTime == 0 {
+		// Only default to a non-zero block time.
+		if blockTime := ctx.BlockTime(); blockTime != zeroTime {
+			ah.BlockTime = blockTime.Unix()
+		}
+	}
+}
+
+// PopulateAction interprets `default:"..."` tags and specially handles an
+// embedded ActionHeader struct.
+func PopulateAction(ctx sdk.Context, action Jsonable) Jsonable {
+	oldRv := reflect.Indirect(reflect.ValueOf(action))
+
+	if oldRv.Kind() != reflect.Struct {
+		return action
+	}
+
+	// Shallow copy to a new value.
+	rv := reflect.Indirect(reflect.New(oldRv.Type()))
+	for i := 0; i < rv.NumField(); i++ {
+		oldField := oldRv.Field(i)
+		field := rv.Field(i)
+		fieldType := rv.Type().Field(i)
+		if !field.CanSet() {
+			continue
+		}
+
+		// Copy from original.
+		field.Set(oldField)
+		defaultTag, _ := fieldType.Tag.Lookup("default")
+
+		// Populate the embedded *ActionHeader struct.
+		if fieldType.Anonymous && fieldType.Type == actionHeaderType {
+			ahp := field.Addr().Interface().(*ActionHeader)
+			setActionHeaderDefaults(ctx, defaultTag, ahp)
```

Comment:
It feels weird to use the `default` tag an an embedded header like this. I know it might be more verbose, but I think a `defaultType` tag would be more appropriate.